### ISSUE TITLE
GenericWriter should write map keys to matching columns

### ISSUE THREAD

**Issue by stoewer** [parquet-go/parquet-go#8]
Given a parquet file with a schema generated from the following model:

```go
type Inner struct {
	FieldB int
	FieldC string
}

type Model struct {
	FieldA string
	Nested Inner  
}
```

The `GenericWriter` should be able to write data from an an alternative model, where `Nested` is represented as map, as long as the map keys match column names from the original schema:

```go
type AltModel struct {
	FieldA string
	Nested map[string]any
}

data := []AltModel{
	{
		FieldA: "a",
		Nested: map[string]any{"FieldB": 11, "FieldC": "c"},
	},
}

schema := parquet.SchemaOf(new(Model))
w := parquet.NewGenericWriter[AltModel](f, schema)
w.Write(data)
```

In its current implementation `GenericWriter` panics for the above code. Here is a [gist that reproduces the error](https://gist.github.com/stoewer/258ed681c0016111b8c8ff618f931351)

This feature is useful when writing data to a schema where parts of the schema (i.e. the struct `Inner`) are defined dynamically at runtime.

**Comment by stoewer** [comment]
Copied from segmentio/parquet-go#510 as this feature is still relevant

**Comment by gernest** [comment]
From the parquet spec this is what it says about Maps 

`MAP` is used to annotate types that should be interpreted as a map from keys
to values. `MAP` must annotate a 3-level structure:

```
<map-repetition> group <name> (MAP) {
  repeated group key_value {
    required <key-type> key;
    <value-repetition> <value-type> value;
  }
}
```

Correct me if I'm wrong. You are basically trying to have  something like 

```
type AltModel struct {
	FieldA string
	Nested map[string]any
}
```

Where you can read/write  on the schema where the value for nested can be anything eg `string|int64` . This is not possible because the spec spec wants a concrete type of the value before hand 

something like 

```
	type AltModel[T string | int64] struct {
		FieldA string
		Nested map[string]T
	}
```

The feature you are requesting is not possible , it goes against the parquet spec. I have a use case of something similar where I store my running binary prometheus metrics and visualise locally, may be useful to you .

This is my model

```go
type Sys struct {
	Timestamp time.Time         `parquet:"timestamp,timestamp,zstd"`
	Labels    map[string]string `parquet:"labels" parquet-key:",dict,zstd" parquet-value:",dict,zstd"`
	Name      string            `parquet:"name,zstd"`
	Value     float64           `parquet:"value,zstd"`
}
```

In my opinion, it is safer to stick to the spec. Also when reading there must not be ambiguity between the type we are  reading . In our case we must never read a `group (MAP)` parquet value from a  plain `group`

**Comment by stoewer** [comment]
Hi @gernest, thanks for having a look at this feature request.
I think there is some kind of misunderstanding here. What I’m asking for has nothing to do with the parquet standard as it does not concern the schema or how the data is stored, it is only about what input is accepted to write to the exact same schema.

What I’m asking for is the following: Given a schema or partial schema matching this struct.

```go
type Model struct {
    FieldA int,
    FieldB string,
}
``` 

It should be possible to write to this schema with a map that has the following content.

```go
map[string]any{
	“FieldA”: 0,
	“FieldB”: “foo”,
}
```

For the input a map with key “FieldA” and an integer value should be treated the same way as a struct field `FieldA int`.

Take this [`write()`](https://gist.github.com/stoewer/258ed681c0016111b8c8ff618f931351#file-parquet_write_map-go-L74) function as an example.

**Comment by gernest** [comment]
Oh! sorry, I completely missed that. I think I understand now.

Basically, this feature is the ability to read `group` into a `map[string]any` where the schema fields/columns are automatically mapped to the map keys and column/field values are mapped to the map values?

Something similar to what json package does with un marshalling   to  `map[string]any`  right?

**Comment by gernest** [comment]
Gentle ping @stoewer @kevinburkesegment ,

So, I gave this some thought. The feature itself is simple to implement, however 
it brings a very big maintenance burden because of the infinity number of requirements
that might arise when using it.

Let me explain by example. I call this an example but its pretty much a full
implementation without tests/error handling. I only cover base types `integers`,
`booleans`, `floats`, `strings` and nested `structs`. We will see that supporting other
data types will be easy based on the api we expose.

## Goal

Given a schema with model

```go
	type Inner struct {
		FieldB string
		FieldC string
	}
	type Model struct {
		FieldA string
		Nested Inner
	}
```

writing a parquet file  with data 

```go
	data := []Model{
		{FieldA: "oneA",
			Nested: Inner{FieldB: "1", FieldC: "oneC"},
		},
		{FieldA: "twoA",
			Nested: Inner{FieldB: "2", FieldC: "twoC"},
		},
		{FieldA: "threeA",
			Nested: Inner{FieldB: "3", FieldC: "threeC"},
		},
		{FieldA: "fourA",
			Nested: Inner{FieldB: "4", FieldC: "fourC"},
		},
		{FieldA: "fiveA",
			Nested: Inner{FieldB: "5", FieldC: "fiveC"},
		},
	}
```

**Expectation**:

We should be a ble to read the file without specifying the `Model` into
something like `[]map[string]any` 

__json__
```json
[
    {
        "FieldA": "oneA",
        "Nested": {
            "FieldB": "1",
            "FieldC": "oneC"
        }
    },
    {
        "FieldA": "twoA",
        "Nested": {
            "FieldB": "2",
            "FieldC": "twoC"
        }
    },
    {
        "FieldA": "threeA",
        "Nested": {
            "FieldB": "3",
            "FieldC": "threeC"
        }
    },
    {
        "FieldA": "fourA",
        "Nested": {
            "FieldB": "4",
            "FieldC": "fourC"
        }
    },
    {
        "FieldA": "fiveA",
        "Nested": {
            "FieldB": "5",
            "FieldC": "fiveC"
        }
    }
]
```


# The code

We take a simple approach. 

1. Open a file
2. Read the file and transform into intermediate representation
3. Convert the intermediate representation into `[]any`

Let's start with step 2 by defining our intermediate representation

```go
// Wraps *parquet.Column. Track relevant column metadata and column values.This
// is linked list of the full schema tree with nested structs as children.
type column struct {
	values []parquet.Value
	// We use this  to transform parquet.Value to a native go value.
	kind parquet.Kind
	// The name of the column. It shows up in the result.
	name string
	// This is important to account for the cases of repeated columns where the
	// number of rows will not be equal to the number of columns.
	//
	// In this implementation repeated columns are not supported, but it is
	// something nice to have.
	rows int64
	// This is nil if this  is a leaf column
	children []*column
}

// Tries to find row count on column c. This is not optimal, but good enough for
// the demo. We don't know row count from the root column as it is not a leaf
// column. This function traverses the children trying to find a leaf with and
// selects the first one it finds.
func (c *column) rowsCount(r *int64) {
	if len(c.children) == 0 {
		// leaf node
		*r = c.rows
		return
	}
	for _, n := range c.children {
		n.rowsCount(r)
	}
}

// transforms the columns into the final
func (c *column) build() (o []any) {
	var rows int64
	c.rowsCount(&rows)
	o = make([]any, rows)
	for i := range o {
		o[i] = c.emit(i)
	}
	return
}

// given row index i constructs a go value from this column. This assumes c is a
// leaf column
func (c *column) buildAt(i int) any {
	switch c.kind {
	case parquet.Boolean:
		return c.values[i].Boolean()
	case parquet.Int32:
		return c.values[i].Int32()
	case parquet.Int64:
		return c.values[i].Int64()
	case parquet.Float:
		return c.values[i].Float()
	case parquet.Double:
		return c.values[i].Double()
	case parquet.ByteArray:
		return c.values[i].String()
	default:
		panic("unsupported type " + c.kind.String())
	}
}

// constructs a go object from row index i. c an be a leaf or group node. For
// leaf node result will be integers | floats | string | boolean and for group
// column result is a map[string]any where the map keys are the children columns
// mapped to their built values
func (c *column) emit(i int) any {
	if len(c.children) == 0 {
		return c.buildAt(i)
	}
	o := make(map[string]any)
	for _, e := range c.children {
		o[e.name] = e.emit(i)
	}
	return o
}
```


 Now we have intermediate representation we can combine step 1 and 2 into a
`read` function


```go

// reads all leaf descendants of root column and transform it into a linked list
// of column.
func read(root *parquet.Column) *column {
	co := &column{
		name: root.Name(),
	}
	if root.Leaf() {
		co.kind = root.Type().Kind()
		pages := root.Pages()
		for {
			page, err := pages.ReadPage()
			if err != nil {
				if errors.Is(err, io.EOF) {
					break
				}
				panic(err)
			}
			v := make([]parquet.Value, page.NumValues())
			page.Values().ReadValues(v)
			co.values = append(co.values, v...)
			co.rows += page.NumRows()
		}
		return co
	}
	for _, col := range root.Columns() {
		co.children = append(co.children, read(col))
	}
	return co
}
```


Now you can combine the three steps to get the desired result. Here is a simple
go program that reads parquet files from stdin and prints pretty json with the
structure intact in stdout.

```go
package main

import (
	"bytes"
	"encoding/json"
	"errors"
	"io"
	"io/ioutil"
	"log"
	"os"

	"github.com/parquet-go/parquet-go"
)

func main() {
	b, err := ioutil.ReadAll(os.Stdin)
	if err != nil {
		log.Fatal(err)
	}
	f, err := parquet.OpenFile(bytes.NewReader(b), int64(len(b)))
	if err != nil {
		log.Fatal(err)
	}
	co := read(f.Root())
	e := json.NewEncoder(os.Stdout)
	e.SetIndent("", " ")
	e.Encode(co.build())
}

// Wraps *parquet.Column. Track relevant column metadata and column values.This
// is linked list of the full schema tree with nested structs as children.
type column struct {
	values []parquet.Value
	// We use this transform parquet.Value to a native go value.
	kind parquet.Kind
	// The name of the column. It shows up in the result.
	name string
	// This is important to account for the cases of repeated columns where the
	// number of rows will not be equal to the number of columns.
	//
	// In this implementation repeated columns are not supported, but it is
	// something nice to have.
	rows int64
	// This is nil if this  is a leaf column
	children []*column
}

// Tries to find row count on column c. This is not optimal, but good enough for
// the demo. We don't know row count from the root column as it is not a leaf
// column. This function traverses the children trying to find a leaf with and
// selects the first one it finds.
func (c *column) rowsCount(r *int64) {
	if len(c.children) == 0 {
		// leaf node
		*r = c.rows
		return
	}
	for _, n := range c.children {
		n.rowsCount(r)
	}
}

// transforms the columns into the final
func (c *column) build() (o []any) {
	var rows int64
	c.rowsCount(&rows)
	o = make([]any, rows)
	for i := range o {
		o[i] = c.emit(i)
	}
	return
}

// given row index i constructs a go value from this column. This assumes c is a
// leaf column
func (c *column) buildAt(i int) any {
	switch c.kind {
	case parquet.Boolean:
		return c.values[i].Boolean()
	case parquet.Int32:
		return c.values[i].Int32()
	case parquet.Int64:
		return c.values[i].Int64()
	case parquet.Float:
		return c.values[i].Float()
	case parquet.Double:
		return c.values[i].Double()
	case parquet.ByteArray:
		return c.values[i].String()
	default:
		panic("unsupported type " + c.kind.String())
	}
}

// constructs a go object from row index i. c an be a leaf or group node. For
// leaf node result will be integers | floats | string | boolean and for group
// column result is a map[string]any where the map keys are the children columns
// mapped to their built values
func (c *column) emit(i int) any {
	if len(c.children) == 0 {
		return c.buildAt(i)
	}
	o := make(map[string]any)
	for _, e := range c.children {
		o[e.name] = e.emit(i)
	}
	return o
}

// reads all leaf descendants of root column and transform it into a linked list
// of column.
func read(root *parquet.Column) *column {
	co := &column{
		name: root.Name(),
	}
	if root.Leaf() {
		co.kind = root.Type().Kind()
		pages := root.Pages()
		for {
			page, err := pages.ReadPage()
			if err != nil {
				if errors.Is(err, io.EOF) {
					break
				}
				panic(err)
			}
			v := make([]parquet.Value, page.NumValues())
			page.Values().ReadValues(v)
			co.values = append(co.values, v...)
			co.rows += page.NumRows()
		}
		return co
	}
	for _, col := range root.Columns() {
		co.children = append(co.children, read(col))
	}
	return co
}
```

Here is my shell when running the generated parquet file from the example you
shared

```
$ cat  paq.parquet |go run tmp/main.go 
[
 {
  "FieldA": "oneA",
  "Nested": {
   "FieldB": "1",
   "FieldC": "oneC"
  }
 },
 {
  "FieldA": "twoA",
  "Nested": {
   "FieldB": "2",
   "FieldC": "twoC"
  }
 },
 {
  "FieldA": "threeA",
  "Nested": {
   "FieldB": "3",
   "FieldC": "threeC"
  }
 },
 {
  "FieldA": "fourA",
  "Nested": {
   "FieldB": "4",
   "FieldC": "fourC"
  }
 },
 {
  "FieldA": "fiveA",
  "Nested": {
   "FieldB": "5",
   "FieldC": "fiveC"
  }
 }
]
```


## NOTES

### The feature is very wasteful

As we have seen of the example. `GenericWriter` is already wasteful.
Maintaining another wasteful abstraction is too much. By wasteful I mean we
need to read into `[]parquet.Value` and convert individual values to respective
native types.

### Why the round trip?

Parquet is columnar, we read and write columns. Reading rows is already a leaky
abstraction but this will take it to the next level. Basically we go from `Row -> Col` then `Col-> Row` again. 

I get that this round trip can be justified, then now we  have a new problem how much to expose , finding the lowest common denominator functionality wise. 

- Do we allow reading subset of the columns ?
- This is limited to full file reading do we change this ? We don't have access to the root column from only row groups or `parquet.Schema`
- Do we accept shape suggestions ? 

### Excessive any and interface{} is a nightmare when things goes wrong

Thanks goodness we  have generics, but as they stand now we can't fully
capitalize on them for saner api.

We can't express something like this 

```go
type Result interface {
	int64 | float64 | bool | string | map[string]Result | []Result
}
```


## Is this useful ?

Yes

## Should this be included in this package ?

I don't think so, it can live separately into a different package. There is no
internal dependency needed to accomplish this. Users will have the freedom to
optimize for their use case.


What do you think ?

**Comment by stoewer** [comment]
Thank you for putting so much thought into this. 

> Basically, this feature is the ability to read group into a map[string]any where the schema fields/columns are automatically mapped to the map keys and column/field values are mapped to the map values?

Yes, the map keys should be mapped to the colums/fields of the schema. However, as stated in the title, this is about *writing* and not *reading* parquet files.

> We should be able to read the file without specifying the Model into something like []map[string]any

I’m not quite sure what you mean by “withoutspecifying the Model”. In my example code snippets and the linked gist, the original model is used to generate the schema for the initialization of the generic writer:

```go
schema := parquet.SchemaOf(new(Model))
w := parquet.NewGenericWriter[AltModel](f, schema)
```

> The feature is very wasteful

Not necessarily. Since the original schema (`Model`) and the alternative model containing the map (`AltModel`) are both present when the `GenericWriter` is created, it should be possible to generate a matching `writeRowsFunc` that looks up the fields in the map and writes them to the destination columns.

I think this could be implemented in the function [ `writeRowsFuncOfMap()`](https://github.com/parquet-go/parquet-go/blob/3d3ee2af9ab7b0fe5457d09867607727882ee2c5/column_buffer_go18.go#L343):

```go
func writeRowsFuncOfMap(t reflect.Type, schema *Schema, path columnPath) writeRowsFunc {
	if _, exists := schema.Lookup(path...); !exists {
		return writeRowsFuncOfNonLeafMap(t, schema, path)
	}

	// … rest of the current code
}

// to be implemented
func writeRowsFuncOfNonLeafMap(t reflect.Type, schema *Schema, path columnPath) writeRowsFunc {
	// return a function that looks up map the map keys for the expected fields under path
	// and writes to the respective columns
}
```

**Comment by gernest** [comment]
Thanks for the feedback.

Either writing/reading. The rationale is still the same. This should be straight forward to implement as a separate library. 

Happy to help you with the starting point implementation for writing  to unblock you.  Unless if I missed something I believe I have addressed this ticket.

**Comment by stoewer** [comment]
> This should be straight forward to implement as a separate library

The approach that I suggested above seems better suitable to me than a separate package. What do you think?

**Comment by gernest** [comment]
> The approach that I suggested above seems better suitable to me than a separate package. What do you think?

Honestly Its something I would have done as well, however I'm wearing a maintainer hat . Which makes me look at this at a different perspective.

- If I allow writing `map[T]any`, I should allow reading `map[T]any`  as well.
- Passing `map[string]any` as `T` raises a lot of questions because we are losing the static aspect of passing `T`. If I was programming in `zig` a lot of `GenericWriter[T]` code will be evaluated at compile time and go away .  While this is simple on the surface dealing with `any` on static `T` context is not somthing that I'm comfortable with.
- I try my best to avoid ambiguity when writing spec compliant code.  Writing as `X` and reading as `Y` isn't  sustainable especially when things stop working.

I might change my mind if we get a different working implementation maybe ?   My biggest concern is the ambiguity, where we have to make assumptions and enforce them when writing/reading.

**Comment by gernest** [comment]
I'm marking this as addressed. Feel free to reopen in case I missed something.
